src/components/application_manager/src/policies/policy_handler.cc

```diff
@@ -317,55 +322,71 @@ bool PolicyHandler::PolicyEnabled() const {
 
 bool PolicyHandler::LoadPolicyLibrary() {
   SDL_LOG_AUTO_TRACE();
-  sync_primitives::AutoWriteLock lock(policy_manager_lock_);
-  if (!PolicyEnabled()) {
-    SDL_LOG_WARN(
-        "System is configured to work without policy "
-        "functionality.");
-    policy_manager_.reset();
-    return false;
-  }
+  ExchangePolicyManager(nullptr);
 
-  if (CreateManager()) {
-    policy_manager_->set_listener(this);
-    event_observer_ = std::shared_ptr<PolicyEventObserver>(
-        new PolicyEventObserver(this, application_manager_.event_dispatcher()));
-  }
+  auto create_policy_manager_instance = [this]() {
+    sync_primitives::AutoWriteLock lock(policy_manager_lock_);
+    if (!PolicyEnabled()) {
+      SDL_LOG_WARN(
+          "System is configured to work without policy "
+          "functionality.");
+      return std::shared_ptr<PolicyManager>();
+    }
 
-  return (policy_manager_.use_count() != 0);
-}
+    void* const dl_policy_handle = dlopen(kLibrary.c_str(), RTLD_LAZY);
 
-bool PolicyHandler::CreateManager() {
-  void* policy_handle = dlopen(kLibrary.c_str(), RTLD_LAZY);
-  const char* error = dlerror();
-  if (!policy_handle) {
-    SDL_LOG_ERROR((error == NULL
-                       ? "Unknown error in dlopen while loading policy table"
-                       : error));
-    return false;
-  }
+    if (!dl_policy_handle) {
+      SDL_LOG_ERROR("An error occurs while calling dlopen");
+      return std::shared_ptr<PolicyManager>();
```

Comment:
@IGapchuk @SNiukalov Is this is a good approach to return shared_ptr as bool here? How do you think?

Reply:
@LitvinenkoIra This place is the scope of the lambda function that returns the `std::shared_ptr<PolicyManager>`.